src/utils/applyChangeToValue.spec.js

```diff
@@ -0,0 +1,180 @@
+import expect from 'expect'
+import applyChangeToValue from './applyChangeToValue'
+import markupToRegex from './markupToRegex'
+
+describe('#applyChangeToValue', () => {
+  const userMarkup = '@[__display__](user:__id__)'
+  const emailMarkup = '@[__display__](email:__id__)'
+  const defaultDisplayTransform = (id, display) => display
+  const config = [
+    {
+      markup: userMarkup,
+      regex: markupToRegex(userMarkup),
+      displayTransform: defaultDisplayTransform,
+    },
+    {
+      markup: emailMarkup,
+      regex: markupToRegex(emailMarkup),
+      displayTransform: defaultDisplayTransform,
+    },
+  ]
+
+  const value =
+    "Hi @[John Doe](user:johndoe), \n\nlet's add @[joe@smoe.com](email:joe@smoe.com) to this conversation..."
+  const plainText =
+    "Hi John Doe, \n\nlet's add joe@smoe.com to this conversation..."
+
+  const displayTransform = id => `<--${id}-->`
+  const plainTextDisplayTransform =
+    "Hi <--johndoe-->, \n\nlet's add <--joe@smoe.com--> to this conversation..."
+
+  it('should correctly add a character at the end, beginning, and in the middle of text', () => {
+    let changed = 'S' + plainText
+    let result = applyChangeToValue(value, config, changed, 0, 0, 1)
+    expect(result).toEqual('S' + value)
+
+    changed = plainText + 'E'
+    result = applyChangeToValue(
+      value,
+      config,
+      changed,
+      plainText.length,
+      plainText.length,
+      changed.length
+    )
+    expect(result).toEqual(value + 'E')
+
+    changed = "Hi John Doe, \n\nlet's Madd joe@smoe.com to this conversation..."
+    result = applyChangeToValue(value, config, changed, 21, 21, 22)
+    expect(result).toEqual(
+      "Hi @[John Doe](user:johndoe), \n\nlet's Madd @[joe@smoe.com](email:joe@smoe.com) to this conversation..."
+    )
+  })
+
+  it('should correctly delete single characters and ranges of selected text', () => {
+    // delete "i"
+    let changed =
+      "H John Doe, \n\nlet's add joe@smoe.com to this conversation..."
+    let result = applyChangeToValue(value, config, changed, 2, 2, 1)
+    expect(result).toEqual(
+      "H @[John Doe](user:johndoe), \n\nlet's add @[joe@smoe.com](email:joe@smoe.com) to this conversation..."
+    )
+
+    // delete "add "
+    changed = "Hi John Doe, \n\nlet's joe@smoe.com to this conversation..."
+    result = applyChangeToValue(
+      value,
+      config,
+      changed,
+      plainText.indexOf('add '),
+      plainText.indexOf('add ') + 'add '.length,
+      plainText.indexOf('add ')
+    )
+    expect(result).toEqual(
+      "Hi @[John Doe](user:johndoe), \n\nlet's @[joe@smoe.com](email:joe@smoe.com) to this conversation..."
+    )
+  })
+
+  it('should correctly add ranges of pasted text and replace the selected range with the new range', () => {
+    // add range
+    let changed = plainText.replace('add', 'add add')
+    let result = applyChangeToValue(
+      value,
+      config,
+      changed,
+      plainText.indexOf('add') + 'add'.length,
```

Comment:
Yup... I definitively think I would better understand this spec when those params are in an object with proper keys.